cpp/src/io/utilities/hostdevice_vector.hpp

```diff
@@ -91,6 +92,12 @@ class hostdevice_vector {
     return reinterpret_cast<T const *>(d_data.data()) + offset;
   }
 
+  operator cudf::detail::device_span<T>() { return {d_data.data(), max_elements}; }
+  operator cudf::detail::device_span<T const>() const { return {d_data.data(), max_elements}; }
+
+  operator cudf::detail::host_span<T>() { return {h_data, max_elements}; }
+  operator cudf::detail::host_span<T const>() const { return {h_data, max_elements}; }
+
```

Comment:
It might also make sense to store the device side data of `hostdevice_vector` in a `rmm::device_uvector` instead of a `device_buffer`. This will probably re-use the `device_span` constructor
```
  template <typename C, std::enable_if_t<is_device_span_supported_container<C>::value>* = nullptr>
  constexpr device_span(C& in) : base(thrust::raw_pointer_cast(in.data()), in.size())
```

Reply:
AFAICT, we would still need the conversion operators to `device_span` here, but we wouldn't have to pass the size separately. I do like this change in general, we should look into polishing `hostdevice_vector` at some point.